<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>document中this被劫持的问题</title>
</head>

<body>
  <button id="button">按钮</button>
</body>
<script>
  // this 被劫持的问题
  var _getElementById = document.getElementById
  document.getElementById = function (id) {
    alert(1)
    return _getElementById(id) // (1)
  }
  var button = document.getElementById('button')

// 执行这段代码，我们看到在弹出 alert(1)之后，紧接着控制台抛出了异常：Uncaught TypeError: Illegal invocation
// 异常发生在(1) 处的_getElementById( id )这句代码上，此时_getElementById 是一个全局函数，
// 当调用一个全局函数时，this 是指向 window 的，而 document.getElementById 方法的内部实现需要使用 this 引用，
// this 在这个方法内预期是指向 document，而不是 window, 这是错误发生的原因。
</script>
<script>
  var _getElementById = document.getElementById
  document.getElementById = function () {
    alert(1)
    // 把 document 当作上下文 this 传入_getElementById：
    return _getElementById.apply(document, arguments)
    // return window._getElementById.apply(document, arguments)
  }
  var button1 = document.getElementById('button')
  console.log(button1)
</script>

<script>
  // 前置装饰
  Function.prototype.before = function (beforefn) {
    var __self = this // 保存原函数的引用
    return function () { // 返回包含了原函数和新函数的"代理"函数
      beforefn.apply(this, arguments) // 执行新函数，且保证 this 不被劫持，新函数接受的参数
      // 也会被原封不动地传入原函数，新函数在原函数之前执行
      return __self.apply(this, arguments) // 执行原函数并返回原函数的执行结果，
      // 并且保证 this 不被劫持
    }
  }

  // 后置装饰
  Function.prototype.after = function (afterfn) {
    var __self = this
    return function () {
      var ret = __self.apply(this, arguments)
      afterfn.apply(this, arguments)
      return ret
    }
  }

  document.getElementById = document.getElementById.before(function () {
    alert(1)
  })
  var button2 = document.getElementById('button')
  console.log(button2)
</script>

</html>